### PR TITLE
include 'lastName' field in 'ContactiblePerson' interface and populate it from API responses

### PR DESCRIPTION
**Pull request by tomsmyers**
## What does this pull request do?

include 'lastName' field in 'ContactiblePerson' interface and populate it from API responses

## What is the intent behind these changes?

we require the PP lastname for emails about changes to the referral.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/HMPPSAuthService.kt [hunk 1]
```diff
@@ -98,15 +100,15 @@ class HMPPSAuthService(
           .retrieve()
           .bodyToMono(UserDetailResponse::class.java)
           .withRetryPolicy()
-          .map { it.name.substringBefore(' ') },
+          .map { Pair(it.name.substringBefore(' '), it.name.substringAfterLast(' ')) },
```
Comment (sldblog) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/HMPPSAuthService.kt:103]: _nit_
Would using `it.name.split` with a limit make this simpler?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/HMPPSAuthService.kt [hunk 2]
```diff
@@ -23,6 +23,7 @@ class UnverifiedEmailException : RuntimeException()
 data class UserDetail(
   override val firstName: String,
   override val email: String,
+  override val lastName: String,
 ) : ContactablePerson
```
Comment (sldblog) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/HMPPSAuthService.kt:27]: _observation_
Given the following interface and classes:

```kt
interface ContactablePerson {
  val firstName: String
  val email: String
  val lastName: String
}

data class UserDetail(
  override val firstName: String,
  override val email: String,
  override val lastName: String,
) : ContactablePerson

data class ResponsibleOfficer(
  val firstName: String?,
  val email: String?,
  val staffId: Long?,
  val lastName: String?,
)

data class ResponsibleProbationPractitioner(
  override val firstName: String,
  override val email: String,
  val deliusStaffId: Long?,
  val authUser: AuthUser?,
  override val lastName: String,
) : ContactablePerson
```

Should `ResponsibleOfficer` and `ResponsibleProbationPractitioner` be merged?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/CaseNotesNotificationsServiceTest.kt [hunk 3]
```diff
@@ -44,9 +44,9 @@ internal class CaseNotesNotificationsServiceTest {
   @Test
   fun `both PPs (responsible officer) and SPs (assignee) get notifications for a sent case note as long as they didn't send it`() {
 
-    whenever(hmppsAuthService.getUserDetail(any<AuthUser>())).thenReturn(UserDetail("sp", "sp@provider.co.uk"))
+    whenever(hmppsAuthService.getUserDetail(any<AuthUser>())).thenReturn(UserDetail("sp", "sp@provider.co.uk", "last"))
```
Comment (sldblog) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/CaseNotesNotificationsServiceTest.kt:47]: _impact of change in other files_

This class changed but its production code equivalent did not -- should we consider extracting these test objects into a constant/factory?

This applies to all tests that only changed in `thenReturn` data and no production code changes

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/CommunityAPIOffenderServiceTest.kt [hunk 4]
```diff
@@ -107,8 +107,8 @@ internal class CommunityAPIOffenderServiceTest {
 
   @Test
   fun `getResponsibleOfficer returns first when there are multiple responsible officers`() {
-    val offenderService = offenderServiceFactory(createMockedRestClient(MockedResponse(offenderManagersLocation, HttpStatus.OK, "[{\"isResponsibleOfficer\": true, \"staff\": {\"forenames\": \"tom\", \"email\": \"tom@tom.tom\"}, \"staffId\": 123}, {\"isResponsibleOfficer\": true}]")))
+    val offenderService = offenderServiceFactory(createMockedRestClient(MockedResponse(offenderManagersLocation, HttpStatus.OK, "[{\"isResponsibleOfficer\": true, \"staff\": {\"forenames\": \"tom\", \"email\": \"tom@tom.tom\", \"surname\": \"jones\"}, \"staffId\": 123}, {\"isResponsibleOfficer\": true}]")))
```
Comment (sldblog) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/CommunityAPIOffenderServiceTest.kt:110]: Given this is an optional field, should we also add a test it maps to `null` if `staff.surname` or `staff` doesn't exist?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/HMPPSAuthService.kt (hunk 1 above)
Comment (tomsmyers) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/HMPPSAuthService.kt:103]: possibly - names could be any number of words e.g. tom david lantos myers, this code just plucks the first and last words out. the alternative would be `it.name.split(' ').first` ... `it.name.split(' ').last` or something - whcih i don't think is any neater. do you?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/CaseNotesNotificationsServiceTest.kt (hunk 3 above)
Comment (tomsmyers) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/CaseNotesNotificationsServiceTest.kt:47]: yeh this is a fair point. we do do this elsewhere with our factory class pattern. i guess in the past it was thought we didn't need a factory for such a skinny object, but as you say, it does make sense for this reason.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/CommunityAPIOffenderServiceTest.kt (hunk 4 above)
Comment (tomsmyers) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/CommunityAPIOffenderServiceTest.kt:110]: yes it would be a fair test. i will add it now.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/HMPPSAuthService.kt (hunk 1 above)
Comment (sldblog) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/HMPPSAuthService.kt:103]: 👍 yeah, this is fine, thanks

**Review by sldblog — APPROVED** [review]
LGTM. Is it worth adding the factory changes as part of this PR?
